index.bs

```diff
@@ -8972,6 +9018,371 @@ worker=] algorithm:
 
+## The storage Module ## {#module-storage}
+
+The <dfn export for=modules>storage</dfn> module contains functionality and
+events related to storage.
+
+A <dfn>storage partition</dfn> is a namespace within which the user agent may
+organize persistent data such as [=cookies=] and local storage.
+
+A <dfn>storage partition key</dfn> uniquely identifies a [=storage partition=].
+
+### Definition ### {#module-storage-definition}
+
+[=Remote end definition=]
+
+<pre class="cddl remote-cddl">
+StorageCommand = (
+  storage.GetCookies //
+  storage.SetCookie //
+  storage.DeleteCookies
+)
+</pre>
+
+[=Local end definition=]
+
+<pre class="cddl local-cddl">
+StorageResult = (
+  storage.GetCookiesResult //
+  storage.SetCookieResult //
+  storage.DeleteCookiesResult
+)
+</pre>
+
+### Types ### {#module-storage-types}
+
+#### The storage.PartitionKey Type #### {#type-storage-PartitionKey}
+
+[=Remote end definition=] and [=Local end definition=]
+
+<pre class="cddl local-cddl remote-cddl">
+storage.PartitionKey {
+    ? userContext: text,
+    ? sourceOrigin: text,
+    Extensible,
+}
+</pre>
+
+The <code>storage.PartitionKey</code> type represents a [=storage partition key=].
+
+A [=remote end=] has a [=/set=] of <dfn>storage partition key attributes</dfn> which contains zero or more implementation-defined string values.
+
+A [=remote end=] has a [=/map=] of <dfn>default values for storage partition key attributes</dfn> which contains zero or more implementation-defined entries. Each key must be a member of the [=storage partition key attributes=] which the [=remote end=] does not require as input to [=storage.getCookies=], [=storage.setCookie=], and [=storage.deleteCookies=]; the corresponding value is a the value which the [=remote end=] uses when one is not provided.
+
+<div algorithm>
+To <dfn>expand a storage partition spec</dfn> given |partition spec| and |cookie domain|:
+
+1. If |partition spec| is a [=browsing context id=]:
+
+  1. Let |browsing context| be the result of [=trying=] to [=get a browsing context=] given |partition spec|.
+
+  1. Let |partition key| be the [=storage partition key|key=] of |browsing context|'s [=associated storage partition=].
+
+  1. Return [=success=] with data |partition key|.
+
+1. Otherwise, if |partition spec| is null:
+
+  1. Let |partition key| be an empty [=/map=].
+
+1. Otherwise:
+
+  1. Assert: |partition spec| is a [=/map=] with zero or more entries.
+
+  1. Let |partition key| be |partition spec|.
+
+1. For each |name| → |default value| in the [=default values for storage partition key attributes=]:
+
+  1. If |partition key|[|name|] does not [=map/exist=]:
+
+    1. [=map/Set=] |partition key|[|name|] to |default value|.
+
+1. If |partition key|["<code>sourceOrigin</code>"] does not [=map/exist=]:
+
+  1. If |cookie domain| is null, then:
+
+    1. Return [=error=] with [=error code=] [=unknown source origin=].
+
+  1. [=map/Set=] |partition key|["<code>sourceOrigin</code>"] to |cookie domain|.
```

Comment:
That's a good point. It seems to me like we have two options:

- in the absence of an explicitly-provided `sourceOrigin` value, use the cookie "domain" attribute as the fallback only if it includes a scheme and port (and return and error if it does not)
- require `sourceOrigin` to be specified

Is that right?

Reply:
So I asked a bit about this, and although I didn't (yet) get a very clear answer it looks like gecko is using (scheme, host, port) as the key (with some extra considerations around not setting cookies for public suffix parts of the domain). It also appears that it has support for the port being unset, but I'm not yet totally sure if that means something other than the default port.

So yes, I think there are a few options:
1. Use the `domain` value where possible. However that (afaict?) doesn't ever have a host and port. So we'd have to pick a default scheme at least. Or maybe we could set cookies for both `http` and `https`, with the default port, although the spec currently doesn't anticipate a 1:many relationship between the request and the actual cookies set. Also many testing setups might use non-default ports for local servers.
2. Error if `sourceOrigin` isn't provided. This has the considerable disadvantage that it makes it easy to write tests that don't work cross-browser.
3. Allow the `domain` value to be a full origin, but only set the actual domain part on the cookie. Although in that case one might wonder whether `domain: "https://example.com:8000"` should also set the `secure` flag. In general making the fields not correspond directly to the actual cookie fields seems likely to cause confusion.
4. Allow an implementation-defined default behaviour (e.g. it could pick a default scheme and port based on sites that have been visited in the current session). But that seems incredibly hard to reason about or make work cross browser.

So unfortunately I think that the best thing for now might be to error out if the source origin isn't supplied and then allow clients to provide a more user friendly API (or maybe during implementation it would become obvious how we could relax the constraint). But I'd be interested to hear what client authors think about this approach.

Reply:
Thanks, @jgraham; I believe [this commit](https://github.com/w3c/webdriver-bidi/pull/501/commits/2315ab086700606905b3e10be466e1bfe1f7a51e) implements your feedback. In this revision, the two partition key attributes we've been discussing (`userContext` and `sourceOrigin`) no longer appear to have any normative requirements at all. Is that because I've misunderstood the intent?

Reply:
I think under the definition of `PartitionKey` you can't just say "keys are implementation defined" but need to actually specify for those keys in the spec what property of the storage partition they actually correspond to, noting that implementations are not required to partition on all keys. 

I also think we should probably require that additional keys are vendor prefixed in until standardised; that would avoid a situation where two browsers implemented a partition with slightly different semantics and it was impossible to specify a key that worked in both.

Reply:
@jgraham I believe that the "sourceOrigin" corresponds to the first item in [the Storage standard's "storage key" tuple](https://storage.spec.whatwg.org/#storage-key), but I'm confused about what makes BiDi's "storage partition key" conceptually distinct from that standard's "storage key." While the "storage key" doesn't [yet](https://privacycg.github.io/storage-partitioning/) include a notion of "user context" (nor is is extensible), I'm suddenly wondering why we shouldn't amend it based on the needs we've identified here. Could you help me understand the interaction between these two concepts?

Reply:
Conceptually they're pretty similar, I think. Although Cookies are a bit special because they're intrinsically keyed on destination domain (and historically _not_ keyed on source origin), whereas other storage types are only keyed on source (i.e. document) origin and don't have the concept of a destination. 

I think it might be easier just to define the concepts we need for BiDi rather than trying to build on other DOM specs; hopefully if we're defining the same underlying reality we converge on the same model and we can eventually share a defintion. But at least for BiDi we're not really expecting anyone to _change_ their storage model, so I think a small amount of duplication is mostly harmless and means we're not so exposed to changes in other standards as they try to unify a DOM-exposed version of how this works.

Reply:
@jgraham If I understand you correctly, this is to be an informative explanation (since normatively defining something like Firefox's Total Cookie Protection seems out of scope for BiDi). That's how I've attempted to integrate your feedback in [the latest commit](https://github.com/w3c/webdriver-bidi/pull/501/commits/4d74327953991bb1866ec9629af4b8061db3d24e) (that also introduces the restrictions on implementation-defined attributes that you recommended). User contexts seems a little more straightforward thanks to gh-570; I've just inserted an issue reference so that this patch can land before that one. Does this look good to you?

I'm also thinking about renaming the term "default values for storage partition key attributes" to "optional storage partition key attributes." The latter would highlight the relationship with the "required [storage] partition key attributes," but on the other hand, it would obscure their difference in type. Any thoughts on that change?